x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlConfigVersion.java

```diff
@@ -0,0 +1,389 @@
+     */
+    public static MlConfigVersion max(MlConfigVersion version1, MlConfigVersion version2) {
+        if (version1.isLegacy != version2.isLegacy) {
+            return version1.isLegacy ? version2 : version1;
+        }
+
+        return version1.id > version2.id ? version1 : version2;
+    }
+
+    /**
+     * Returns {@code true} if the specified version is compatible with this running version of Elasticsearch.
+     */
+    public static boolean isCompatible(MlConfigVersion version) {
+        return version.onOrAfter(MINIMUM_COMPATIBLE);
+    }
+
+    public boolean after(MlConfigVersion version) {
+        if (this.isLegacy != version.isLegacy) {
+            return this.isLegacy ? false : true;
+        }
+        return version.id < id;
+    }
+
+    public boolean onOrAfter(MlConfigVersion version) {
+        if (this.isLegacy != version.isLegacy) {
+            return this.isLegacy ? false : true;
+        }
+        return version.id <= id;
+    }
+
+    public boolean before(MlConfigVersion version) {
+        if (this.isLegacy != version.isLegacy) {
+            return this.isLegacy ? true : false;
+        }
+        return version.id > id;
+    }
+
+    public boolean onOrBefore(MlConfigVersion version) {
+        if (this.isLegacy != version.isLegacy) {
+            return this.isLegacy ? true : false;
+        }
+        return version.id >= id;
+    }
+
+    public boolean between(MlConfigVersion lowerInclusive, MlConfigVersion upperExclusive) {
+        if (upperExclusive.onOrBefore(lowerInclusive)) throw new IllegalArgumentException();
+        return onOrAfter(lowerInclusive) && before(upperExclusive);
+    }
+
+    public static MlConfigVersion fromVersion(Version version) {
+        return fromId(version.id);
+    }
+
+    public static Version toVersion(MlConfigVersion mlConfigVersion) {
+        if (mlConfigVersion.id < MINIMUM_COMPATIBLE.id) {
+            throw new IllegalArgumentException("Cannot convert " + mlConfigVersion.id + ". Incompatible version");
+        }
+        return Version.fromId(mlConfigVersion.id);
+    }
+
+    public static MlConfigVersion getMinMlConfigVersion(DiscoveryNodes nodes, String mlConfigVersionNodeAttrName) {
+        return getMinMaxMlConfigVersion(nodes, mlConfigVersionNodeAttrName).v1();
+    }
+
+    public static MlConfigVersion getMaxMlConfigVersion(DiscoveryNodes nodes, String mlConfigVersionNodeAttrName) {
+        return getMinMaxMlConfigVersion(nodes, mlConfigVersionNodeAttrName).v2();
+    }
+
+    public static Tuple<MlConfigVersion, MlConfigVersion> getMinMaxMlConfigVersion(
+        DiscoveryNodes nodes,
+        String mlConfigVersionNodeAttrName
+    ) {
+        MlConfigVersion minMlConfigVersion = MlConfigVersion.CURRENT;
+        MlConfigVersion maxMlConfigVersion = MlConfigVersion.MINIMUM_COMPATIBLE;
+        for (DiscoveryNode node : nodes) {
+            String mlConfigVersionStr = node.getAttributes().get(mlConfigVersionNodeAttrName);
```

Comment:
For nodes that don't have the attribute we should call `node.getVersion()` and infer the ML config version from that.

As this code stands it will convert a `null` attribute to the current version. This is guaranteed to be wrong. The version of the node that didn't have the attribute must be old.